[1.95.0] Document provenance and its use in const eval (#707)
ferrocene/specification: src/glossary.rst

:t:`[token]s` and produces a stream of tokens, and is invoked via attribute
:c:`derive`.

derived pointer
^^^^^^^^^^^^^^^

> A :dt:`derived pointer` is a :t:`pointer` obtained by :t:`borrowing`, copying a :t:`pointer`, loading a stored :t:`pointer`, performing :t:`pointer` arithmetic, or :t:`casting` a :t:`pointer`.

Reviewer: does "loading" mean reading

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/fls/707/51b787946f8e3f5cbbf5df632e1c50f31f8d9ed6..bed3e05da1c9acd5db8850be5b60fbaa2a2c1a6b)*

Author: It most likely involves a read of the actual address, but also of the pointer metadata (if any).

Reviewer: let's use "read" instead

An :dt:`indirection type` is a :t:`type` whose :t:`[value]s` refer to memory
locations.

indirection value
^^^^^^^^^^^^^^^^^

> An :dt:`indirection value` is either a :t:`raw pointer` or a :t:`reference`.

Reviewer: section 4.7, Indirection Types, includes function pointers... should they be included here

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/fls/707/3c73fb2263ce3ffc9f7d7592a910d42372cad001..21bb6de29dd62a21d6af72afbf60ee74adf4c82c)*

Author: No, as [indicated](https://rust-lang.zulipchat.com/#narrow/channel/520710-t-lang.2Ffls/topic/.E2.9C.94.20Provenance/with/607336254) by bjorn3.

A :dt:`program entry point` is a :t:`function` that is invoked at the start of
a Rust program.

provenance
^^^^^^^^^^

> :dt:`Provenance` is an optional property of :t:`[pointer]s` that restricts the memory locations the :t:`pointer` may access, the timespan during which the accesses may occur, and whether the accesses may read from or write to memory.

Reviewer: - I understand "access" as ability to read from the memory locations, such that a read is implied (unless you find the suggestion too vague)
- "write to memory" does not indicate which memory
```suggestion
:dt:`Provenance` is an optional property of :t:`[pointer]s` that restricts the memory locations the :t:`pointer` may access, the timespan during which the accesses may occur, and whether the :t:`pointer` can write to such memory locations.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/fls/707/f7b0d8703918dba4cbf11344f7d48119fcd75871..979328b4c2722253a8b61385af3aa69d78d50431)*

Author: Note that you left out reads! This is important, as provenance dictates **read** and **write** permissions.

Reviewer: access sounds like it implies reads, at least in the context of a programming language

Author: I do not agree. "Memory access" has always meant reads and writes. You have RAM (Random **Access** Memory), which is both read and written, as opposed to ROM (Read-**only** Memory).

Also, from `core::ptr`:

Whether a pointer is valid depends on the operation it is used for (read or write), and the extent of the memory that is **accessed** (i.e., how many bytes are **read/written**)

If you do not like the verb "to access", then you will have to suggest a better alternative.

Reviewer: If access means read and/or write, then we are here saying: "restricts the memory locations the pointer may read-write, when those reads-writes can happen, and whether those are reads and/or writes".

That is, "access" implies reads, but only includes writes optionally (hence my suggestion).

An alternative, in case that still doesn't work (to avoid "access'):

> Provenance is an optional property of pointers that restricts the memory locations the pointer may read, the timespan during which the read may occur, and whether the pointer can also write to such memory locations.

It's a little vague, so maybe:

> Provenance is an optional property of pointers that restricts the memory locations the pointer may read, the timespan during which the read may occur, and whether those reads can also include writes.

Not sure. What do you think.

Author: More like "and whether the reads and/or writes are allowed in the first place".

I do not think your suggestion is correct, as once again you are leaving out writes from several of the provenance properties. I am not sure why you are trying to "save on text", and cause the rule to imply additional semantics. One of the goals of the FLS is to spell out rules clearly, even if there is repetition, so that there is no doubt, implication, or ambiguity.

Basically "provenance" says
- You are allowed to access memory locations from 0x000A to 0x001A.
- You are allowed to access those memory locations from this part of your program to this part of your program.
- You are allowed to read and/or write from/to those memory locations.

There is "write-only" memory from the point of view of your program! For example, GPU memory can behave this way. Your program simply dumps data onto the GPU buffers, and the GPU reads it. Your program does not read from the GPU.

Reviewer: > There is "write-only" memory from the point of view of your program! For example, GPU memory can behave this way.

good to know

See :s:`WeakKeyword`.

well-formed indirection value
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> A :dt:`well-formed indirection value` is an :t:`indirection value` with :t:`provenance`, where all bytes that comprise the :t:`indirection value` are part of the same group.

Reviewer: the definition feels incomplete, but I don't know yet how to fill it... maybe inspiration can be taken from [a review from my old pr](https://github.com/rust-lang/fls/pull/695/changes#r3481039037)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/fls/707/3c73fb2263ce3ffc9f7d7592a910d42372cad001..21bb6de29dd62a21d6af72afbf60ee74adf4c82c)*

Author: Maybe
```rst
... bytes that comprise the :t:`indirection value` are initialized and are part
of the same group.
```
Basically we want to say that an indirection value does not (somehow) contain the bytes of something else.

Reviewer: is this about "group of bytes"

Author: Right.

I think what the Reference PRs and `core::ptr` are trying to say is that you can obtain a value by doing sketchy things, such as memory transmutation. If you plan on using that value as an indirection value with provenance, then the value better have "correct bytes", what 2091 calls "bytes must be fragments of the same original pointer value in the correct order" and 2138 calls "part of an adjacent group of correctly-ordered bytes that form an entire pointer".

Author: I updated the definition of "well-formed indirection value".

Author: I updated the PR with the terminology suggested by Ralph.

Reviewer 2: The update to pointers makes sense along the lines of what Ralf suggested, cool.

I see two possible gaps remaining in the definition (details and rustc examples on the `src/values.rst` thread for `fls_ffh8mAkebORJ`):

1. It still requires provenance, so provenance-free pointer constants that rustc accepts (`core::ptr::null()`, `5 as *const u8`) are still ruled out. The fragment restriction only concerns bytes that carry provenance.
2. "Fragments of the same :t:`original pointer`" over-constrains the source: fragments of a *derived* pointer (e.g. a cast) reassemble fine; the compiler only needs the bytes to rebuild a single pointer value with provenance.

For this entry:

```suggestion
A :dt:`well-formed pointer` is a :t:`pointer` where either no byte of the
:t:`pointer` carries :t:`provenance`, or every byte of the :t:`pointer` is the
corresponding byte of a single :t:`pointer` with :t:`provenance`.
```

("initialized" is fine to keep or drop — it's already implied by `fls_6lg0oaaopc26`.)

Author: Incorporated.
